trin-core/src/portalnet/overlay_service.rs

```diff
@@ -522,6 +530,7 @@ impl<TContentKey: OverlayContentKey + Send> OverlayService<TContentKey> {
             "[{:?}] Handling ping request from node={}. Ping={:?}",
             self.protocol, source, request
         );
+        self.metrics.report_inbound_ping();
```

Comment:
The noop pattern was inspired by metrics setup in [trinity](https://github.com/ethereum/trinity/blob/eaa3b040ffdf0848b8e00a5f329480662ecc7c11/trinity/components/builtin/metrics/instruments.py). I think I see an advantage to both approaches. Using `Option` allows us to remove some of the code. On the other hand, I feel `Noop` does a better job of encapsulating logic and simplifying the code in `overlay_service`. 

Aka `self.metrics.report_inbound_ping();` is better to read than `self.metrics.and_then(|m| m.report_inbound_ping());`.

Metrics have very little, if anything, to do with the actual functionality happening in `overlay_service` - any code complexity in `overlay_service` should be limited to the `OverlayService` functionality. Aka - if we can abstract away any metrics logic into its own module, we should. 

It's also worth mentioning, that this is just the very beginning of the metrics reporting. It's hard for me to predict exactly what the metrics reporting code will end up looking like (honestly, it'll probably look very different), since we'll be reporting all kinds of metrics from all over the codebase. Keeping this in mind, it makes sense to me to keep the metrics code as modular / logically separated as possible.  

> If we do decide to keep the MetricsReporter trait, then it would make more sense to add another generic type parameter with trait bounds to the overlay instead of a trait object here. This is because each overlay would only ever use a single concrete type that implements MetricReporter.

You know how I feel about generics & traits in rust. However misguided I might be. Adding another generic type to `OverlayService`, especially for something as tbd as our metrics reporting system, just feels messy. 

On this point however, I'd be curious to hear your pushback. My main problem with generic traits is that they're messy and difficult to read / parse, and typically result in some added complexity when dealing with nested types. This is why I tend to avoid using them, unless there's a direct advantage. Maybe you can help point out the flaws in my reasoning? The argument that it's just more "Rust-y" is not super convincing tbh - but I could be missing something.

Ok - that's a lot of rambling, ultimately, I'd just like to make sure that whatever we go with, it leaves us flexible down the road as the metrics code evolves and might end up moving around the codebase. Curious to hear your thoughts

Reply:
I'll solely address the first issue since, like you said, the second issue may not exist after we resolve the first.

I've seen the no-op pattern before, but personally it doesn't make sense to me to define and maintain some extra structure that does nothing, especially given the prevalence of `Option` in Rust. If you have an argument in favor of it, then I'm open to it.

Also, I don't think the `and_then` syntax adds any significant friction to readability. If readability is an issue, we can use `if let Some(...)`.

Given that we are in the early stages of metrics reporting, I am in favor of only using a single concrete struct to handle that functionality, so that we do not "pre-abstract" things before we know what shape that abstract thing should take.

If we want to take steps toward something more general, we probably don't want a struct (or trait) that defines methods like `report_inbound_ping`, but instead something like `increment_counter(counter: CounterMetric)`, where we initialize the struct with a set of metrics to register and manage.

Reply:
I see both points. I think that for now, the `metrics` field being an optional `Metrics` struct is the best bet, and then switch to a trait if we define multiple different metrics reporters down the line, but I don't love the `NoOp` pattern. I think that a `None` signaling no metrics is more explicit and understandable. Also I agree that updating the NoOp implementation anytime we update the metrics trait sounds annoying.

I definitely think that adding a generic type parameter to the class would be overkill.